state/machine.go

```diff
@@ -2218,40 +2255,90 @@ func (m *Machine) SetUpgradeSeriesStatus(unitName string, status model.UnitSerie
 		if err := m.isStillAlive(); err != nil {
 			return nil, errors.Trace(err)
 		}
-		coll, closer := m.st.db().GetCollection(machineUpgradeSeriesLocksC)
-		defer closer()
-		var lock upgradeSeriesLockDoc
-		err := coll.FindId(m.Id()).One(&lock)
+		docIndex, err := m.getUnitIndex(unitName, status)
 		if err != nil {
-			return nil, errors.BadRequestf("machine %q is not locked for upgrade", m)
-		}
-		docIndex := -1
-		for i, unitStatus := range lock.PrepareUnits {
-			if unitStatus.Id == unitName {
-				// short circuit if there is nothing to do
-				if unitStatus.Status == status {
-					return nil, jujutxn.ErrNoOperations
-				}
-				docIndex = i
-			}
-		}
-		if docIndex == -1 {
-			return nil, fmt.Errorf("cannot get upgrade series lock for unit %q of machine %v: %v", unitName, m.Id(), err)
+			return nil, errors.Trace(err)
 		}
-		logger.Debugf("Requested unit name: %q, Unit indexed %q", unitName, lock.PrepareUnits[docIndex].Id)
-
 		return setUpgradeSeriesTxnOps(m.doc.Id, unitName, docIndex, status, bson.Now()), nil
 	}
 	err := m.st.db().Run(buildTxn)
 	if err != nil {
 		err = onAbort(err, ErrDead)
-		logger.Errorf("cannot set series upgrade status for unit %q of machine %q: %v", unitName, m.Id(), err)
 		return err
 	}
+	return nil
+}
 
+// SetUpgradeSeriesStatus sets the machine status of a series upgrade.
+func (m *Machine) SetMachineUpgradeSeriesStatus(status model.MachineSeriesUpgradeStatus) error {
+	buildTxn := func(attempt int) ([]txn.Op, error) {
+		if attempt > 0 {
+			if err := m.Refresh(); err != nil {
+				return nil, errors.Trace(err)
+			}
+		}
+		if err := m.isStillAlive(); err != nil {
+			return nil, errors.Trace(err)
+		}
+		statusSet, err := m.isMachineUpgradeSeriesStatusSet(status)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if statusSet {
+			return nil, jujutxn.ErrNoOperations
+		}
+		return setMachineUpgradeSeriesTxnOps(m.doc.Id, status), nil
+	}
+	err := m.st.db().Run(buildTxn)
+	if err != nil {
+		err = onAbort(err, ErrDead)
+		return err
+	}
 	return nil
 }
 
+func (m *Machine) isMachineUpgradeSeriesStatusSet(status model.MachineSeriesUpgradeStatus) (bool, error) {
+	lock, err := m.getUpgradeSeriesLock()
+	if err != nil {
+		return false, err
+	}
+	return lock.PrepareStatus == status, nil
+}
+
+func setMachineUpgradeSeriesTxnOps(machineDocID string, status model.MachineSeriesUpgradeStatus) []txn.Op {
+	return []txn.Op{
+		{
+			C:      machinesC,
+			Id:     machineDocID,
+			Assert: isAliveDoc,
+		},
+		{
+			C:      machineUpgradeSeriesLocksC,
+			Id:     machineDocID,
+			Update: bson.D{{"$set", bson.D{{"prepare-status", status}}}},
+		},
+	}
+}
+
+func completeUpgradeSeriesTxnOps(machineDocID string, units []unitStatus) []txn.Op {
```

Comment:
the code would be easier to read if this was directly below CompleteUpgradeSeries()

Reply:
I would rather keep them grouped.